Add ADF41513 Driver

rodrigo455 opened the pull request:
## PR Description

- This PR is not meant to be merged, as it targets an upstream mirror
- I want to collect feedback and run CI pipelines if possible
- This PR adds the ADF41513 driver implementation and documentation
- ADF41513 is a dual modulus PLL frequency synth
- Current implementation is based on existing PLL drivers
- the driver implements a clock provider/supplier, but clock-scales is not taken care here, because that feature is not available in this branch

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have compiled my changes, including the documentation
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly
- [ ] I have provided links for the relevant upstream lore

gastmaier commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1196 @@
+	struct adf41513_platform_data *pdata;
+	struct iio_dev *indio_dev;
+	struct adf41513_state *st;
+	struct clk *ref_clk;
```
```suggestion
	struct clk *ref_clk = NULL;
```
if (pdata->clkin), ref_clk is never set.

gastmaier commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1196 @@
+	if (st->ref_clk)
+		st->ref_freq_hz = clk_get_rate(st->ref_clk);
+	else
+		st->ref_freq_hz = pdata->clkin;
```
```suggestion
	if (pdata->clkin)
		st->ref_freq_hz = pdata->clkin;
	else
		st->ref_freq_hz = clk_get_rate(st->ref_clk);
```

flows a little more naturally IMO, since ref_clk is set if clkin is not set

artursartamonovsadi commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1196 @@
+	struct mutex			lock;
+
+	/* Cached register values */
+	u32				regs[14];
```
array size is 14, can there array size used as define? Later on the REG0 and REG13 used in for loop. could be from MAX_REG number to 0. Assume no more the 14 registers are used?

artursartamonovsadi commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1196 @@
+	8910, 9720, 10530, 11340, 12150, 12960
+};
+
+static int adf41513_parse_uhz(const char *str, u64 *freq_uhz)
```
can this be replaced with kstrtou64
from kernels include/linux/kstrtox.h

nunojsa commented on include/linux/iio/frequency/adf41513.h:
```diff
@@ -0,0 +1,264 @@
+ * @load_enable_sync:		Enables load enable sync with reference input.
+ * @freq_resolution_uhz:	Target frequency resolution in uHz (default = 1 Hz).
+ */
+struct adf41513_platform_data {
```
I see this is still a draft but I might just tell it now. Most likely you don't really need the header as I don't expecte anyone other than the driver itself to consume this. So, just move it to the source file.

Another comment is for `struct adf41513_platform_data`. Name it differently (like adf41513_state or adf41513_data). platform data is a thing from the past :)

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1196 @@
+			return PTR_ERR(pdata);
+	} else {
+		pdata = spi->dev.platform_data;
+		if (!pdata) {
```
As I mentioned, drop the platform data stuff. We just want to support FW properties.

nunojsa reviewed (commented):
Just some comments about using legacy platform_data. Once you mark this ready for reviewing I'll give a more in depth review.

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+    description: Power supply for the device (3.3V)
+
+  chip-enable-gpios:
+    description: |
```
drop `|`. Same in all other places

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+    $ref: /schemas/types.yaml#/definitions/uint64
+    description:
+      If set, the PLL tunes to this frequency (in Hz) on driver probe.
+      Range is 1 GHz to 26.5 GHz for ADF41513, and 1 GHz to 10 GHz for ADF41510.
```
If there is a range, add proper constrains in here with minimum and maximun. That said, we might need to better explain this property and why we need it during probe. Seems something that can be done by userspace at startup

rodrigo455 replied:
thinking of it, I will make this a required option. During init, we need to program all registers (according to the user guide), so we need to set an initial frequency. It could be a pre-defined value, but such frequency may depend on onboard RF filters designed for specific ranges.

nunojsa replied:
Is it something that we could somehow infer at probe? Just asking because it might be a valid question upstream? A property like this is very uncommon in DT (not to speak about being mandatory) so I would at least give a few words about it either in the description (if you can avoid driver details) or in the commit message.

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+    maximum: 7200
+    description:
+      Charge pump current in microamps. The value will be rounded to the
+      nearest supported value. Default is 2400 µA.
```
If there's a default, add it in the property (not in description).

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+      8: Readback
+      10: CLK1 divider output
+      13: R divider/2 output
+      14: N divider/2 output
```
You might get asked to turn this into strings

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+    description:
+      External charge pump resistor value in ohms. This sets the maximum
+      charge pump current along with the charge pump current setting.
+      Default is 2700 ohms.
```
For the above two there are standard unit suffixes that you should use:

https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/property-units.yaml

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,265 @@
+      edge of the reference signal to prevent glitches.
+
+  adi,freq-resolution-uhz:
+    $ref: /schemas/types.yaml#/definitions/uint64
```
I guess we need this because the default suffix goes to uint32?

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+		if (st->regs_hw[i] != st->regs[i] ||
+		    sync_mask & (1 << i)) {
+			st->buf = cpu_to_be32(st->regs[i] | i);
+			ret = spi_write(st->spi, &st->buf, 4);
```
sizeof(st->buf)

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+			    ADF41513_MAX_INT_8_9 : ADF41513_MAX_INT_4_5;
+
+	u64 freq_error_uhz;
+	u16 int_value = (u16)div64_u64_rem(result->target_frequency_uhz,
```
no need for the cast

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+	.debugfs_reg_access = &adf41513_reg_access,
+};
+
+#ifdef CONFIG_OF
```
should not need the above

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+#endif
+}
+
+static int adf41513_parse_dt(struct adf41513_state *st)
```
You are using FW agnostic APIs, so I would rename it to adf41513_parse_fw()

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+	}
+
+	/* configure IIO device */
+	indio_dev->name = st->data.name[0] ? st->data.name : "adf41513";
```
The name should be constant. Just use chip_info struct kind of thing if you need more than the name. Or just a const char pointer passed through of_id and spi_id tables.

nunojsa reviewed (commented):
Ok, I'm done for now :). We'll need some iterations on this one and skipped more questionable/complicated things for now. We are also pushing the limits of what can go into a DT property so be ready to justify them when upstreaming.

rodrigo455 commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+	 */
+	_ADF41513_EXT_INFO("frequency", ADF41513_FREQ),
+	_ADF41513_EXT_INFO("frequency_resolution", ADF41513_FREQ_RESOLUTION),
+	_ADF41513_EXT_INFO("refin_frequency", ADF41513_FREQ_REFIN),
```
those attributes are pretty much the same available in other PLL drivers.

nunojsa replied:
Then we might be asked for a generic ABI file like sys-bus-iio-frequency

rodrigo455 committed:
ADF41513/ADF41510 PLL frequency synthesizers

This patch series adds support for the Analog Devices ADF41513 and ADF41510
ultralow noise PLL frequency synthesizers. These devices are designed for
implementing local oscillators (LOs) in high-frequency applications.
The ADF41513 covers frequencies from 1 GHz to 26.5 GHz, while the ADF41510
operates from 1 GHz to 10 GHz.

Key features supported by this driver:
- Integer-N and fractional-N operation modes
- High maximum PFD frequency (250 MHz integer-N, 125 MHz fractional-N)
- 25-bit fixed modulus or 49-bit variable modulus fractional modes
- Digital lock detect functionality
- Phase resync capability for consistent output phase
- Load Enable vs Reference signal syncronization

The series includes:
1. PLL driver implementation
2. Device tree bindings documentation
3. IIO ABI documentation

# Describe the purpose of this series. The information you put here
# will be used by the project maintainer to make a decision whether
# your patches should be reviewed, and in what priority order. Please be
# very detailed and link to any relevant discussions or sites that the
# maintainer can review to better understand your proposed changes. If you
# only have a single patch in your series, the contents of the cover
# letter will be appended to the "under-the-cut" portion of the patch.

# Lines starting with # will be removed from the cover letter. You can
# use them to add notes or reminders to yourself. If you want to use
# markdown headers in your cover letter, start the line with ">#".

# You can add trailers to the cover letter. Any email addresses found in
# these trailers will be added to the addresses specified/generated
# during the b4 send stage. You can also run "b4 prep --auto-to-cc" to
# auto-populate the To: and Cc: trailers based on the code being
# modified.

To: linux-kernel@vger.kernel.org
To: linux-iio@vger.kernel.org
To: devicetree@vger.kernel.org
To: linux-doc@vger.kernel.org
Cc: Jonathan Cameron <jic23@kernel.org>
Cc: David Lechner <dlechner@baylibre.com>
Cc: Nuno Sá <nuno.sa@analog.com>
Cc: Andy Shevchenko <andy@kernel.org>
Cc: Lars-Peter Clausen <lars@metafoo.de>
Cc: Michael Hennerich <Michael.Hennerich@analog.com>
Cc: Rob Herring <robh@kernel.org>
Cc: Krzysztof Kozlowski <krzk+dt@kernel.org>
Cc: Conor Dooley <conor+dt@kernel.org>
Cc: Jonathan Corbet <corbet@lwn.net>
Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

---
Changes in v4:
- Proper usage of units.h macros
- Simplifications to DT property parsing
- Adjustments to return value handling
- Drop of simple DT property node example
- Link to v3: https://lore.kernel.org/r/20260108-adf41513-iio-driver-v3-0-23d1371aef48@analog.com

Changes in v3:
- Use FIELD_MODIFY macro in driver implementation
- Drop refin_frequency iio attribute
- Drop muxout-select property from dt-bindings (and rename logic-level property)
- Use -mhz suffix in power-up frequency property
- Address documentation issues
- Link to v2: https://lore.kernel.org/r/20251219-adf41513-iio-driver-v2-0-be29a83d5793@analog.com

Changes in v2:
- separate driver implementation from extra features and improve commit messages
- use macros from units.h
- explanation of custom parse function: adf41513_parse_uhz
- reorganize driver data structures
- drop clock framework support for now
- reorganize documentation
- Link to v1: https://lore.kernel.org/r/20251110-adf41513-iio-driver-v1-0-2df8be0fdc6e@analog.com

--- b4-submit-tracking ---
# This section is used internally by b4 prep for tracking purposes.
{
  "series": {
    "revision": 4,
    "change-id": "20251110-adf41513-iio-driver-aaca8a7f808e",
    "prefixes": [],
    "history": {
      "v1": [
        "20251110-adf41513-iio-driver-v1-0-2df8be0fdc6e@analog.com"
      ],
      "v2": [
        "20251219-adf41513-iio-driver-v2-0-be29a83d5793@analog.com"
      ],
      "v3": [
        "20260108-adf41513-iio-driver-v3-0-23d1371aef48@analog.com"
      ]
    }
  }
}

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+/* REG2 Bit Definitions */
+#define ADF41513_REG2_PHASE_VAL_MSK		GENMASK_U32(15, 4)
+#define ADF41513_REG2_PHASE_VAL(x)		FIELD_PREP(ADF41513_REG2_PHASE_VAL_MSK, x)
+#define ADF41513_REG2_PHASE_ADJ_MSK		BIT_U32(31)
```
i get it but normally no one really cares about that and use the generic BIT() and GENMASK() macros

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1458 @@
+#define ADF41513_REG12_LOGIC_LEVEL_MSK		BIT_U32(27)
+#define ADF41513_REG12_LOGIC_LEVEL(x)		FIELD_PREP(ADF41513_REG12_LOGIC_LEVEL_MSK, x)
+#define ADF41513_REG12_MUXOUT_MSK		GENMASK_U32(31, 28)
+#define ADF41513_REG12_MUXOUT(x)		FIELD_PREP(ADF41513_REG12_MUXOUT_MSK, x)
```
Still not addressed

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+		st->settings.actual_frequency_uhz +=
+			mul_u64_u64_div_u64(st->settings.frac1,
+					    st->settings.pfd_frequency_uhz,
+					    ADF41513_FIXED_MODULUS);
```
Alignment on open parenthesis. Ditto for all other places

rodrigo455 replied:
is there a handy command using uncrustify or clang-format to address those?

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+	result->ref_doubler = st->data.ref_doubler_en ? 1 : 0;
+
+	if (st->data.ref_doubler_en &&
+	    st->ref_freq_hz > ADF41513_MAX_REF_FREQ_DOUBLER) {
```
Checkpatch should warn about the above so in this case I guess we can break the 80 col limit as it helps readability

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+	mod2 = min_t(u64,
+		     DIV64_U64_ROUND_CLOSEST(result->pfd_frequency_uhz * 2,
+					     st->data.freq_resolution_uhz * ADF41513_FIXED_MODULUS),
+		     ADF41513_MAX_MOD2);
```
Again hard to read code and also not very pleasant to look at (to be honest :)). Also, I think that even in our tree we have the improved min(). So I thin there's no reason for min_t()

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+};
+MODULE_DEVICE_TABLE(of, adf41513_of_match);
+
+DEFINE_SIMPLE_DEV_PM_OPS(adf41513_pm_ops,
```
static

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+	return 0;
+}
+
+static void adf41513_power_down(void *data)
```
Is this only called as the action handler? If so, no need for locking.

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1438 @@
+	memset(st->regs_hw, 0xFF, sizeof(st->regs_hw));
+
+	/* assume DLD pin is used for digital lock detect */
+	st->regs[ADF41513_REG5] |= ADF41513_REG5_DLD_MODES(ADF41513_DLD_DIG_LD);
```
If it is the first write on that field, dont do |

nunojsa reviewed (commented):
Here it goes my second iteration. As I said, this one will take some...

machschmitt reviewed (commented):
Hi @rodrigo455 ,

Some comments from my side.

rodrigo455 commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1523 @@
+		return ret;
+
+	/* apply computed results to state pll settings */
+	pfd_change = (st->settings.pfd_frequency_uhz != result.pfd_frequency_uhz);
```
that would not be the case. the pfd_frequency is the frequency of the phase detector and it is different from the requested synthetized frequency. For the range of frequencies the system was the designed for, it is likely that the pfd frequency will not change, as it seems to be a project constraint when following guidelines from https://www.analog.com/en/lp/resources/adisimpll.html

the driver allows for pfd frequency change, when the requested frequency is not able to be synthetized due to the supported integer range.

rodrigo455 committed:
dt-bindings: iio: frequency: add adf41513

DT-bindings for ADF41513, an ultralow noise PLL frequency synthesizer that
can be used to implement local oscillators (LOs) as high as 26.5 GHz.
Most properties are based upon an existing PLL device properties
(e.g. ADF4350).

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

nunojsa commented on drivers/iio/frequency/Kconfig:
```diff
@@ -29,6 +29,17 @@ endmenu
+config ADF41513
+	tristate "Analog Devices ADF41513 PLL Frequency Synthesizer"
+	depends on SPI
+	depends on COMMON_CLK
```
The above one is a bit questionable. I guess we have it because of the clock provider feature right? But I would assume that feature to be optional?

rodrigo455 replied:
will remove it

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1526 @@
+
+	ret = of_clk_add_provider(spi->dev.of_node, of_clk_src_simple_get, clk);
+	if (ret)
+		return ret;
```
The above is using deprecated APIs. See:

https://elixir.bootlin.com/linux/v6.18-rc7/source/drivers/clk/clk.c#L4640

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1526 @@
+	struct adf41513_state *st = to_adf41513_state(hw);
+
+	guard(mutex)(&st->lock);
+	return (unsigned long)div_u64(adf41513_pll_get_rate(st), MICROHZ_PER_HZ);
```
I think the cast is already implicit

nunojsa reviewed (commented):
Some more comments from. I have ran out of time (public holiday for me today :)).

FWIW, I think we are getting ready for upstream v2

rodrigo455 committed:
iio: frequency: adf41513: driver implementation

The driver is based on existing PLL drivers in the IIO subsystem and
implements the following key features:

- Integer-N and fractional-N (fixed/variable modulus) synthesis modes
- High-resolution frequency calculations using microhertz (µHz) precision
  to handle sub-Hz resolution across multi-GHz frequency ranges
- IIO debugfs interface for direct register access
- FW property parsing from devicetree including charge pump settings,
  reference path configuration and muxout options
- Power management support with suspend/resume callbacks
- Lock detect GPIO monitoring

The driver uses 64-bit microhertz values throughout PLL calculations to
maintain precision when working with frequencies that exceed 32-bit Hz
representation while requiring fractional Hz resolution.

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

rodrigo455 committed:
iio: frequency: adf41513: features on frequency change

Set Bleed current when PFD frequency changes (bleed enabled when in
fractional mode). Set lock detector window size, handling bias and
precision. Add phase resync support, setting clock dividers when
PFD frequency changes.

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

rodrigo455 commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,232 @@
+      Charge pump current (I_CP) in microamps. The value will be rounded to the
+      nearest supported value. Range of acceptable values depends on the
+      charge pump resistor value, such that 810 mV <= I_CP * R_SET <= 12960 mV.
+      This value depends on the loop filter design.
```
no... the constraints here depends on the value that was set for the resistance.

machschmitt commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+
+static const struct iio_chan_spec_ext_info adf41513_ext_info[] = {
+	/*
+	 * Ideally we use IIO_CHAN_INFO_FREQUENCY, but there are
```
nitpicking: "Ideally we use ..." -> "Ideally we would use ..."

machschmitt commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	if (ret >= 0)
+		st->data.muxout_select = ret;
+	else
+		st->data.muxout_select = ADF41513_MUXOUT_TRISTATE;
```
Can be slightly simplified with
```c
	st->data.muxout_select = ret >= 0 ? ret : ADF41513_MUXOUT_TRISTATE;
```

machschmitt commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -1125,6 +1130,148 @@ static const struct iio_info adf41513_info = {
+				    unsigned long *parent_rate)
+{
+	return rate;
+}
```
This looks a bit surprising. If the full 1 GHz to 26.5 GHz range is supported, shouldn't we at least do something like `clamp(rate, ADF41513_MIN_RF_FREQ, ADF41513_MAX_RF_FREQ);`? Does REFin impact the achievable output frequencies? Might also want to clamp differently if ADF41510. Or maybe drop round_rate() if it gets too complicated. I don't think round_rate() is a required clock operation.

rodrigo455 replied:
dropped clk framework support for now

machschmitt reviewed (commented):
Hi @rodrigo455 ,

Here goes my second round of review.
Besides nitpicking on small things, I think the IIO documentation (not the ABI) looks a bit bloated. Some sections/subsections don't add much or talk about things that are documented somewhere else in the kernel. Though, I think the doc would make a great wiki page at ADI wiki or at ADI github as it is. Also, I'd have the ABI doc and IIO driver doc in separate patches. "docs: iio: ..." is not an accurate path for the ABI part of the patch.

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1526 @@
+		if (tmp >= ADF41513_MIN_CP_VOLTAGE_mV && tmp <= ADF41513_MAX_CP_VOLTAGE_mV)
+			st->data.charge_pump_voltage_mv = tmp;
+		else
+			return dev_err_probe(dev, -ERANGE, "I_CP %u uA (%u Ohms) out of range\n",
```
Yes, that is one of the hurdles with `in_range()`. And I have as you suggested in past with a _len kind of macro. I was just stating some more nitpicky reviewers might ask for it :). I personally don't care much

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf41513.yaml:
```diff
@@ -0,0 +1,232 @@
+      reference. The value specifies the resync period in nanoseconds, used
+      to configure clock dividers with respect to the PFD frequency. This value
+      should be set to a value that is at least as long as the worst case lock
+      time, i.e., it depends mostly on the loop filter design.
```
Hmm I'm not 100% convinced with the above. Is the resync_period_ns still meaningful if we change the pfd_frequency_uhz?  I mean, from the datasheet:

`tSYNC = CLK1 × CLK2 × tPFD`

right? So, if we change tPFD, the value we set in the bindings no longer holds true right? Or am I missing something? (note I still did not looked into the current series so I might be completely wrong).

Other thing that comes to mind... Do we know the CLK1 and CLK2 values in the driver? If so, we could just have a boolean property for the resync feature and do the math at runtime?

rodrigo455 replied:
> Is the resync_period_ns still meaningful if we change the pfd_frequency_uhz?

that is why the function that adjusts the clk divs is only called when there is a PFD frequency change.

> So, if we change tPFD, the value we set in the bindings no longer holds true right?

the clk divs are adjusted so that the binding holds true. `t_sync` being the worst locking time.

> Do we know the CLK1 and CLK2 values in the driver? If so, we could just have a boolean property for the resync feature and do the math at runtime?

right now, phase resync is enabled when resync_period_ns is non-zero, and CLK1 and CLK2 are recomputed when there is a PFD change.

nunojsa commented on Documentation/ABI/testing/sysfs-bus-iio-frequency-adf41513:
```diff
@@ -0,0 +1,21 @@
+		Sets the value for Lock Detector count of a PLL, which determines the
+		number of consecutive phase detector cycles that must be within the lock
+		detector window before lock is declared. Lower values increase the lock
+		detection sensitivity.
```
Not 100% sure about this one. Could very well be a DT property. Up to you how you want to try it upstream

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+
+	ret = devm_add_action_or_reset(&spi->dev, adf41513_power_down, st);
+	if (ret)
+		return dev_err_probe(&spi->dev, ret, "Failed to add power down action: %d\n", ret);
```
Given that the above depends on the enable gpio, we could only add the action in case the gpio is available. Then you can assume that in the callback. But more importantly, let's only add resources to the device when it makes sense to do so. Not a big deal though...

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+
+	ret = adf41513_setup(st);
+	if (ret < 0)
+		return dev_err_probe(&spi->dev, ret, "failed to setup device: %d\n", ret);
```
I personally prefer to do any meaningful logs inside the function and just return from the caller. But up to you

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+
+	ret = devm_mutex_init(&spi->dev, &st->lock);
+	if (ret)
+		return dev_err_probe(&spi->dev, ret, "failed to initialize mutex: %d\n", ret);
```
I would expect that if we fail in devm_mutex_init() we will already get lots of logs (basically because we could not allocate memory). So no sure the above adds much.

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	st->data.charge_pump_voltage_mv = ADF41513_DEFAULT_CP_VOLTAGE_mV;
+	ret = device_property_read_u32(dev, "adi,charge-pump-current-microamp", &cp_current);
+	if (!ret) {
+		tmp = (cp_current * cp_resistance) / 1000; /* convert to mV */
```
Andy will likely nitpick about using units.h as well :)

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+		st->regs[ADF41513_REG2] |= ADF41513_REG2_PHASE_ADJ_MSK;
+		st->regs[ADF41513_REG2] &= ~ADF41513_REG2_PHASE_VAL_MSK;
+		st->regs[ADF41513_REG2] |= FIELD_PREP(ADF41513_REG2_PHASE_VAL_MSK, phase_val);
+		return adf41513_sync_config(st, ADF41513_SYNC_REG0);
```
I would prefer to see the above in an helper function. Also looks like we're missing locking

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	/*
+	 * Ideally we use IIO_CHAN_INFO_FREQUENCY, but there are
+	 * values > 2^32 with sub-Hz resolution
+	 */
```
This also needs a better explanation. Kind of we need plus micro (with long long). Something like that

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1526 @@
+		break;
+	default:
+		ret = -EINVAL;
+		break;
```
not addressed

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+		if (readin < ADF41513_LD_COUNT_FAST_MIN || readin > ADF41513_LD_COUNT_MAX ||
+		    !is_power_of_2(readin)) {
+			ret = -EINVAL;
+			break;
```
In cases like this just do `return -EINVAL`

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	switch ((u32)private) {
+	case ADF41513_FREQ_REFIN:
+		st->ref_freq_hz = clk_get_rate(st->ref_clk);
+		val = st->ref_freq_hz;
```
Agreed! It would also be consistent with ADF41513_LOCK_DETECT_COUNT_AVAILABLE

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -673,9 +684,82 @@ static int adf41513_calc_pll_settings(struct adf41513_state *st,
+						st->data.phase_resync_period_ns,
+						1ULL * MICRO * NANO);
+		clk1_div = clamp(int_sqrt(total_div), 1,
+				 ADF41513_MAX_CLK_DIVIDER);
```
The above definitely needs a comment and explanation. What's the assumption for it? Do we assume clk1 = clk2?

rodrigo455 committed:
docs: iio: add documentation for adf41513 driver

Add documentation for ADF41513 driver which describes the device
driver files and shows how userspace may consume the ABI for various
tasks

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

rodrigo455 commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	u32 regs_hw[ADF41513_REG_NUM];
+
+	/* PLL data */
+	struct adf41513_data data;
```
remove the comment, the fact that this is PLL data is kinda obvious

rodrigo455 commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1282 @@
+	_ADF41513_EXT_UHZ_INFO("frequency", ADF41513_FREQ),
+	_ADF41513_EXT_UHZ_INFO("frequency_resolution", ADF41513_FREQ_RESOLUTION),
+	_ADF41513_EXT_UHZ_INFO("pfd_frequency", ADF41513_FREQ_PFD),
+	_ADF41513_EXT_INFO("refin_frequency", ADF41513_FREQ_REFIN),
```
added to the docs. thanks.

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1244 @@
+
+	adf41513_suspend(st);
+	if (st->chip_enable)
+		gpiod_set_value_cansleep(st->chip_enable, 0);
```
Just curious... Are disabling different blocks with the enable_gpio when compared to `adf41513_suspend()`? IOW, do we need both?

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1244 @@
+			return dev_err_probe(dev, -ERANGE,
+					     "power-up frequency %llu Hz out of range\n",
+					     st->data.power_up_frequency_hz);
+	}
```
I though you had the above as mandatory now?

rodrigo455 replied:
yeah, decided to make optional with 10 GHz as default. Still, this value should be set for specific hardware designs so that the initialization sequence starts from a valid configuration.

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1244 @@
+		dev_err(&st->spi->dev, "RF frequency %llu uHz out of range [%llu, %llu] uHz\n",
+			rf_out_uhz, min_rf_freq_uhz, max_rf_freq_uhz);
+		return -EINVAL;
+	}
```
nit: Blank line here :)

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1244 @@
+{
+	u16 max_int = (st->chip_info->has_prescaler_8_9) ?
+		      ADF41513_MAX_INT_8_9 : ADF41513_MAX_INT_4_5;
+
```
blank line not needed. Take care for similar places

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -662,9 +673,84 @@ static int adf41513_calc_pll_settings(struct adf41513_state *st,
+
+	bleed_value *= FIELD_GET(ADF41513_REG5_CP_CURRENT_MSK, st->regs[ADF41513_REG5]) + 1;
+	bleed_value = mul_u64_u64_div_u64(st->settings.pfd_frequency_uhz,
+					  (u64)bleed_value,
```
no need for the cast

nunojsa reviewed (commented):
Ok, it looks like we're getting closer to an upstream v2. My main doubt right now is phase_resync computation

rodrigo455 commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -662,9 +673,84 @@ static int adf41513_calc_pll_settings(struct adf41513_state *st,
+					1ULL * MICRO * NANO);
+	clk1_div = clamp(int_sqrt(total_div), 1,
+			 ADF41513_MAX_CLK_DIVIDER);
+	clk2_div = clamp(DIV_ROUND_CLOSEST(total_div, clk1_div), 1,
```
in previous parts (e.g. [ADF4350](https://www.analog.com/media/en/technical-documentation/data-sheets/ADF4350.pdf)) we would not have clk2, the clk2 divider was derived from the modulus value instead.
I suppose that the digital design is reusing the same module, but exposing a second custom divider.
The datasheet does not favor one or another, using both allows for a wider configuration range

rodrigo455 commented on Documentation/ABI/testing/sysfs-bus-iio-frequency-adf41513:
```diff
@@ -0,0 +1,28 @@
+		frequency accordingly.
+		The value written has no effect until out_altvoltageY_frequency
+		is updated. Consider to use out_altvoltageY_powerdown to power down the
+		PLL and its RFOut buffers during REFin changes.
```
> The value written has no effect until out_altvoltageY_frequency is updated

I can handle this part.

> Consider to use out_altvoltageY_powerdown to power down the PLL and its RFOut buffers during REFin changes

this is not really necessary but avoid glitches in the output. There is an option for the part to synchronize the Latch Enable with the reference signal to avoid that, but I have not enabled that for now, mostly because it forces you to have the reference signal always available, and that is not the case when you are using the PLL as a frequency tracker rather than a frequency synthesizer. I will add that option as a DT property.

rodrigo455 committed:
iio: frequency: adf41513: handle LE synchronization feature

When LE sync is enabled, it is must be set after powering up and must be
disabled when powering down. It is recommended when using the PLL as
a frequency synthesizer, where reference signal will always be present
while the device is being configured.

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1251 @@
+{
+	struct iio_dev *indio_dev;
+	struct adf41513_state *st;
+
```
nit: no need for the blank line

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -0,0 +1,1251 @@
+#include <linux/log2.h>
+#include <linux/math64.h>
+#include <linux/module.h>
+#include <linux/mod_devicetable.h>
```
I think mod_devicetable.h is before module.h

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -737,8 +738,19 @@ static int adf41513_suspend(struct adf41513_state *st)
+	if (st->data.le_sync_en) {
+		st->regs[ADF41513_REG12] |= ADF41513_REG12_LE_SELECT_MSK;
+		ret = adf41513_sync_config(st, ADF41513_SYNC_DIFF);
+	}
```
You say it must disabled when powering down. I would expect to see the line

`st->regs[ADF41513_REG12] &= ~ADF41513_REG12_LE_SELECT_MSK;`

only on this patch

nunojsa commented on drivers/iio/frequency/adf41513.c:
```diff
@@ -663,9 +674,86 @@ static int adf41513_calc_pll_settings(struct adf41513_state *st,
+	clk1_div = clamp(int_sqrt(total_div), 1,
+			 ADF41513_MAX_CLK_DIVIDER);
+	clk2_div = clamp(DIV_ROUND_CLOSEST(total_div, clk1_div), 1,
+			 ADF41513_MAX_CLK_DIVIDER);
```
I'm still wondering why we have the above. If we assume they are the same why not directly writing clk1_div in both register? Or just clk2_div = clk1_div with a comment.

rodrigo455 replied:
when you have:
$$total_{div} = f_{PFD} * t_{sync}$$
$$clk1_{div} = \sqrt{total_{div}}$$
$$clk2_{div} = total_{div}/clk1_{div}$$
you get a closer approximation to the desired value.

nunojsa reviewed (commented):
Ok, take my comments and I think we are good for v2.

rodrigo455 committed:
Documentation: ABI: testing: add common ABI file for iio/frequency

Add ABI documentation file for PLL/DDS devices with frequency_resolution
sysfs entry attribute used by ADF4350 and ADF41513

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>

rodrigo455 committed:
units: Add HZ_PER_GHZ definition

New PLL IIO driver supports output frequency of several GHz.
The new define helps to constraint DT properties and frequency
calculation parameters.

Signed-off-by: Rodrigo Alencar <rodrigo.alencar@analog.com>
